bpo-32309: Implement asyncio.ThreadPool

aeros opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->

Implements `asyncio.ThreadPool`, a high-level asynchronous context manager for concurrently running IO-bound functions without blocking the event loop. The initial implementation relies significantly upon `loop.run_in_executor()`, but the long-term goal is to eventually use a fully native asyncio threadpool (which doesn't rely on `concurrent.futures.ThreadPoolExecutor`). See the bpo issue for more details.

<!-- issue-number: [bpo-32309](https://bugs.python.org/issue32309) -->
https://bugs.python.org/issue32309
<!-- /issue-number -->

Note: Disregard the name of the branch. It has all of the intended changes, but it seems that I accidentally added the changes on top of  branch from my recently merged PR (https://github.com/python/cpython/pull/18057). This has no consequence though other than a misleading branch name, so I'll leave it as is.

aeros committed:
Implement asyncio.ThreadPool

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

aeros committed:
Add whatsnew for 3.9

aeros committed:
Fix asyncio.ThreadPool examples

tirkarthi commented on Lib/test/test_asyncio/test_pools.py:
```diff
@@ -0,0 +1,182 @@
+            self.loop.run_until_complete(pool_cm())
+
+
+class ThreadPoolTests(unittest.TestCase):
```
Could the tests benefit from `IsolatedAsyncioTestCase`? I guess it would remove `run_until_complete` calls along with using `async def main` . Doc : https://docs.python.org/3/library/unittest.html#unittest.IsolatedAsyncioTestCase

aeros replied:
Yeah, that seems like it would be quite useful for reducing a decent chunk of the boilerplate. I'll test it locally when I get the chance later, assuming @1st1 and @asvetlov are on board with using `IsolatedAsyncioTestCase` for these tests.

aeros replied:
Based on my local experimentation so far, it seems be to working as intended and does provide a bit of a readability benefit in terms to reducing white noise. Just waiting on a +1 from @1st1 since this would be the first CPython regression test to use `IsolatedAsyncioTestCase` (outside of `unittest` of course).

aeros replied:
Update: I'd like to consider using `IsolatedAsyncioTestCase` in a separate PR, after the initial version is implemented since we're a bit short on time with the upcoming feature deadline for 3.9 beta.

aeros committed:
Fix asyncio.ThreadPool example

aeros committed:
Improve tearDown() in test_pools

pitrou commented on Doc/library/asyncio-pools.rst:
```diff
@@ -0,0 +1,95 @@
+   ``None``, the amount of threads used will scale based on the
+   number of processors.
+
+   .. coroutinemethod:: run(/, func, *args, **kwargs)
```
Is there a reason to deviate from the `Executor` API (which calls this method `submit`)? Do you fear people might use one for another?

aeros replied:
I don't personally have especially strong views on using `run()` vs `submit()`, @1st1 initially suggested the usage of `run()` being the primary means of executing functions within the thread pool (in the BPO issue). But, it does seem to be more aligned with the existing `run_in_executor()` that this was based on.

Perhaps Yury could elaborate on why he initially chose `run()` over `submit()` as the method name?

tmewett replied:
For additional reference, another pool class, `multiprocessing.pool.Pool`, uses `apply` as a name.

1st1 replied:
```suggestion
   .. coroutinemethod:: run(func, /, *args, **kwargs)
```

pitrou commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,166 @@
+    """Abstract base class for asynchronous pools."""
+
+    async def astart(self):
+        raise NotImplementedError
```
Is it possible to decorate this method with `abc.abstractmethod`, so that the base class can't be instantiated?

aeros replied:
Yeah, that seems like a good idea and would probably be more explicit. The only reason that I didn't use it was just that I modeled it to be similar to the existing abstract classes in asyncio, such as `AbstractServer` and `AbstractEventLoopPolicy`.

But, I see no reason not to add `abc.abstractmethod`. I can update it accordingly.

aeros replied:
Done.

pitrou commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,166 @@
+
+    def __init__(self, concurrency=None):
+        if concurrency is None:
+            concurrency = min(32, (os.cpu_count() or 1) + 4)
```
This heuristic looks a bit gratuitous. I think it would be less surprising for the user to have `concurrency = os.cpu_count()` as a default, just like `concurrent.futures` does.

aeros replied:
@pitrou 

> I think it would be less surprising for the user to have concurrency = os.cpu_count() as a default, just like concurrent.futures does.

ProcessPoolExecutor uses `os.cpu_count()` as its default, but ThreadPoolExecutor uses `min(32, (os.cpu_count() or 1) + 4)` (which is what I based the above on):

https://github.com/python/cpython/blob/02a4d57263a9846de35b0db12763ff9e7326f62c/Lib/concurrent/futures/thread.py#L135-L143

pitrou replied:
Uh, I see. That's the unfortunate result of wanting a default that's "good" for both IO concurrency and parallel execution of CPU-bound tasks...

aeros replied:
Also to clarify, it looks like that was changed fairly recently (last year) as a part of https://bugs.python.org/issue35279. Previously, ThreadPoolExecutor used `(os.cpu_count() or 1) * 5`. 

Since @asvetlov and @methane were involved in that change (https://github.com/python/cpython/pull/13618), do either of you have any thoughts regarding the default max threads for asyncio.ThreadPool? Would the current default for ThreadPoolExecutor also make sense here, or should we potentially consider a different value?

methane replied:
Discussion in bpo-35279 can be applied to this thireadpool too.  Using same default makes sense.
But why do you calculate the default value here?  Why don't you pass None to ThreadPoolExecutor?

aeros replied:
> But why do you calculate the default value here? Why don't you pass None to ThreadPoolExecutor?

The reason the maximum is calculated locally is because a primary goal for `asyncio.ThreadPool` is to transition into a native implementation that doesn't rely on `concurrent.futures`, see the discussion in bpo-32309 for more details. So, while we could just pass None to *max_workers* when it's not specified by the user in this specific version, I think it makes more sense to define the maximum locally for the long term.

pitrou reviewed (commented):
I'll be interested to see the non-wasteful native implementation when you'll get to it :-)

aeros committed:
Co-authored-by: Antoine Pitrou <antoine@python.org>

Use abc.ABC and abc.abstractmethod to prevent direct instantiation
of AbstractPool.

aeros committed:
Fix whitespace

tmewett commented on Doc/library/asyncio-pools.rst:
```diff
@@ -0,0 +1,95 @@
+      Raises a :exc:`RuntimeError` if the threadpool is already running or
+      if it's been closed.
+
+   .. coroutinemethod:: aclose()
```
The naming of `astart` and `aclose` don't seem too clear to me. Would it be more descriptive to name them `start` and `shutdown` (like the concurrent.futures.Executor) or maybe `close` (like multiprocess.pool.Pool)?

aeros replied:
The reasoning behind using `astart` and `aclose` is to clearly communicate that the methods are non-blocking and coroutines that should be awaited. As `start` and `close` are common names in similar APIs, it would be easy to forget that. It also fits quite nicely with the corresponding dunder/system method names of `__aenter__` and `__aexit__`, which is used for async context managers. In most situations though, users should not have to use `astart` and `aclose` explicitly; so I don't think it's an overly significant concern.

tmewett commented on Doc/library/asyncio-pools.rst:
```diff
@@ -0,0 +1,95 @@
+
+.. class:: ThreadPool(concurrency=None)
+
+   An asynchronous threadpool that provides methods to concurrently
```
If I may make a very minor point, it seems "thread pool" with the space is a much more common spelling of this term

aeros replied:
Good point. In retrospect, I believe that I've also seen the spelling "thread pool" far more commonly in comparison to "threadpool". So, I'll use that spelling for the docs instead.

aeros replied:
Done.

aeros committed:
Replace "threadpool" with "thread pool"

Co-authored-by: Tom M <tom@collider.in>

1st1 commented on Doc/library/asyncio-pools.rst:
```diff
@@ -0,0 +1,95 @@
+
+   .. coroutinemethod:: astart()
+
+      Schedule the start of the thread pool and spawn its threads. Note that
```
```suggestion
      Start the thread pool and spawn its threads. Note that
```

With how coroutines are typically (with an `await` expression) it's not about "scheduling an action" it's about "doing an action". I admit the details can be blurry for some people, but to make it simpler I suggest to simply describe what coroutines are doing without mentioning "scheduling"

aeros replied:
Okay, that seems reasonable to me. The use of "scheduling" was mainly there as a way to describe it as "begin this operation, but without blocking". 

However, that might be needlessly over-complicating the documentation. The fact that it's a coroutine method instead of a subroutine method should hopefully explain that to some degree. So, I'll change it as suggested.

1st1 commented on Doc/library/asyncio-pools.rst:
```diff
@@ -0,0 +1,95 @@
+
+   .. coroutinemethod:: aclose()
+
+      Schedule the closing of the thread pool. Note that this function is
```
ditto

1st1 commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,173 @@
+        await self.aclose()
+
+    async def run(self, /, func, *args, **kwargs):
+        if not self._running:
```
```suggestion
    async def run(self, func, /, *args, **kwargs):
        if not self._running:
```

You want `func` to be a valid keyword argument of the function.

aeros replied:
Thanks, good catch. I had not initially considered that *func* might be a valid kwarg for the function being passed to `run()`.

1st1 commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,173 @@
+            raise RuntimeError("threadpool is closed")
+
+        future = self._loop.create_future()
+        thread = threading.Thread(target=self._do_spawn, args=(future,))
```
I'm not sure we need this complication, why don't we call `concurrent.futures.ThreadPoolExecutor()` here directly? Yes, it'll block for a slightly longer amount of time, but IMO that's totally OK. The implementation would be simpler.

aeros replied:
@1st1 
> I'm not sure we need this complication, why don't we call concurrent.futures.ThreadPoolExecutor() here directly? Yes, it'll block for a slightly longer amount of time, but IMO that's totally OK.

The current implementation for `astart()` and `aclose()` is based on the conversation in the bpo issue, mainly starting from [here](https://bugs.python.org/issue32309#msg355862):

> Thus btw makes me think that initializing threads/processes in `__init__` is a bad idea, as spawning can be asynchronous.

In the case of `astart()`, I don't think making it block would be a huge deal in most situations, as the main bottleneck would be waiting for the threads to spawn. This could be a concern when using a large number of threads (depending on the eventual native implementation of `asyncio.ThreadPool`), but otherwise I don't think it would be a real concern for the majority of users. Especially since `ThreadPoolExecutor` spawns threads on-demand as needed, rather than all of them at startup.

In the case of `aclose()`, I think that calling `executor.shutdown()` by itself would be rather detrimental, especially when using `wait=True` (which is the default). This would mean that the event loop would be effectively blocked while the executor is waiting on pending futures to be finished, no?

Instead, with the current implementation, it would allow the coroutine calling `await aclose()` (or exiting the `async with` scope) to suspend while the executor is shutting down in a separate thread until the awaited future is done, allowing tasks or other coroutines to complete work in the meantime. This could be quite impactful to performance, and would allow for more effective multi-tasking.

Thus, I think we could reasonable change `astart()` to directly initialize `ThreadPoolExecutor()` for simplification, but I strongly think that the current `aclose()` implementation should remain for the above reasons. Also, note that the current implementation for `aclose()` almost exactly mirrors what we used for `loop.shutdown_default_executor()`.

1st1 replied:
Ah, you're right. Let's keep the current implementation for both `astart` and `aclose`

methane replied:
How about using `loop.run_in_executor` instead of creating new thread only for starting/closing new threadpool?

We can assume loop.default_executor is ThreadPoolExecutor because:

```
    def set_default_executor(self, executor):
        if not isinstance(executor, concurrent.futures.ThreadPoolExecutor):
            warnings.warn(
                'Using the default executor that is not an instance of '
                'ThreadPoolExecutor is deprecated and will be prohibited '
                'in Python 3.9',
                DeprecationWarning, 2)
        self._default_executor = executor
```

1st1 replied:
> How about using loop.run_in_executor instead of creating new thread only for starting/closing new threadpool?

+1.

aeros replied:
That's a great idea @methane, thanks for the suggestion.

1st1 commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,173 @@
+        self._running = False
+
+        future = self._loop.create_future()
+        thread = threading.Thread(target=self._do_shutdown, args=(future,))
```
ditto

aeros replied:
See https://github.com/python/cpython/pull/18410#discussion_r424869543.

1st1 reviewed (commented):
This looks good, great job Kyle.  IMO this is OK to ship in 3.9. Please address the review comments. @asvetlov what do you think?

aeros committed:
Simplify astart() and aclose() docs

Co-authored-by: Yury Selivanov <yury@edgedb.com>

aeros committed:
Allow *func* to be a kwarg for functions passed to run()

Co-authored-by: Yury Selivanov <yury@edgedb.com>

1st1 commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,173 @@
+            self._closed = True
+            self._loop.call_soon_threadsafe(future.set_result, None)
+        except Exception as ex:
+            self._loop.call_soon_threadsafe(future.exception, ex)
```
```suggestion
            self._loop.call_soon_threadsafe(future.set_exception, ex)
```

1st1 replied:
Would be cool to figure out if we can test this.

aeros replied:
Thanks, good catch. I don't frequently use the low-level API for `asyncio.Future` in my own code, so I forgot that it was `set_exception` instead of `exception`. :-)

@1st1 
> Would be cool to figure out if we can test this.

Do you have something specific in mind, such as ensuring that exceptions are correctly propagated to the result of `await run()`? That should be doable and fairly straightforward. It would make sense IMO to add a test for that.

Otherwise, it might be a bit difficult to directly ensure the exception was correctly set to the future without creating a separate mocked version to use for the tests, since the future is internal to its respective method.

Edit: see https://github.com/python/cpython/pull/18410#issuecomment-628424120.

1st1 commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,173 @@
+            self._running = True
+            self._loop.call_soon_threadsafe(future.set_result, None)
+        except Exception as ex:
+            self._loop.call_soon_threadsafe(future.exception, ex)
```
```suggestion
            self._loop.call_soon_threadsafe(future.set_exception, ex)
```

aeros committed:
Apply suggestions from code review

Update run() docs function signature and use future.set_exception() instead of future.exception()

Co-authored-by: Yury Selivanov <yury@edgedb.com>

aeros committed:
Add test_run_exception()

aeros committed:
Use run_in_executor() for startup/shutdown instead of seperate thread

Co-authored-by: Inada Naoki <songofacandy@gmail.com>

aeros committed:
Add missing conversions from 'threadpool' to 'thread pool'

aeros committed:
Add shutdown_default_executor() to tearDown()

1st1 commented on Lib/asyncio/pools.py:
```diff
@@ -0,0 +1,152 @@
+    async def __aexit__(self, exc_type, exc_value, exc_traceback):
+        await self.aclose()
+
+    async def run(self, /, func, *args, **kwargs):
```
```suggestion
    async def run(self, func, /, *args, **kwargs):
```

This still isn't fixed.

aeros replied:
Sorry, I thought I directly committed that suggestion to the PR using the GitHub UI. I'll fix it manually.

Edit: Never mind, I see the earlier suggestion was for `AbstractPool.run()`. I'll just directly commit this suggestion, thanks.

aeros committed:
Fix ThreadPool.run() function signature

Co-authored-by: Yury Selivanov <yury@edgedb.com>

aeros committed:
Add shutdown_default_executor to ThreadPoolTests.tearDown()
